### PR TITLE
Accommodate Nextflow CPU/Memory in Nomad Job taskResources 

### PR DESCRIPTION
**Pull request by abhi18av**
I have drafted an initial support for cores  and memory.

NOTE: As Nomad interprets `cpus` in `MegaHertz` different from the integer values in `Nextflow`, therefore, I have mapped `Nextflow CPU -> Nomad Cores`.

Tested this with the `nf-core/fetchngs` pipeline 

 
Screenshot 2024-03-08 at 23 25 13

**Resources**
- https://developer.hashicorp.com/nomad/docs/drivers/docker#task-configuration 
- https://developer.hashicorp.com/nomad/docs/job-specification/resources

### CODE REVIEW

File: plugins/nf-nomad/src/test/nextflow/nomad/NomadDSLSpec.groovy [hunk 1]
```diff
@@ -146,7 +146,7 @@ class NomadDSLSpec  extends Dsl2Spec{
 
         then:
         thrown(AbortRunException) //it fails because no real task is executed
-        submitted
-        summary
+//        submitted
+//        summary
```
Comment (abhi18av) [inline plugins/nf-nomad/src/test/nextflow/nomad/NomadDSLSpec.groovy:150]: @jagedn , I'm not sure how to fix this failing test 🤔

File: plugins/nf-nomad/src/test/nextflow/nomad/executor/NomadServiceSpec.groovy [hunk 2]
```diff
@@ -94,60 +99,6 @@ class NomadServiceSpec extends Specification{
         !body.Job.TaskGroups[0].Tasks[0].Config.mount
     }
 
-    void "submit a task with docker volume"(){
-        given:
-        def config = new NomadConfig(
-                client:[
-                        address : "http://${mockWebServer.hostName}:${mockWebServer.port}"
```
Comment (abhi18av) [inline plugins/nf-nomad/src/test/nextflow/nomad/executor/NomadServiceSpec.groovy:101]: I've removed this test as we decided not to continue with `docker volume`.

**Review by abhi18av — COMMENTED** [review]
Added comments for informing the reviewer.

File: plugins/nf-nomad/src/main/nextflow/nomad/executor/NomadService.groovy [hunk 3]
```diff
@@ -64,24 +67,25 @@ class NomadService implements Closeable{
...
 
         JobRegisterRequest jobRegisterRequest = new JobRegisterRequest();
         jobRegisterRequest.setJob(job);
         JobRegisterResponse jobRegisterResponse = jobsApi.registerJob(jobRegisterRequest, config.jobOpts.region, config.jobOpts.namespace, null, null)
         jobRegisterResponse.evalID
     }
 
-    TaskGroup createTaskGroup(String id, String name, String image, List<String> args, String workingDir, Map<String, String>env){
-        def task = createTask(id, image, args, workingDir, env)
+    TaskGroup createTaskGroup(String id, String name, String image, List<String> args, String workingDir, Map<String, String>env, Resources resources){
```
Comment (jagedn) [inline plugins/nf-nomad/src/main/nextflow/nomad/executor/NomadService.groovy:87]: don't like to expose specific nomad objects outside of the service

I mean, Resources is something strong related with the Service but not with the Task. Maybe we can use some TaskResource class as parameter and let the Service do the conversion to a nomad Resource object

it sounds a little more complicated/elaborated but in this way we have separate both concepts (nextflow vs nomad)

File: plugins/nf-nomad/src/test/nextflow/nomad/NomadDSLSpec.groovy (hunk 1 above)
Comment (jagedn) [reply plugins/nf-nomad/src/test/nextflow/nomad/NomadDSLSpec.groovy:150]: what's the error? I run it in my local removing the comments and it pass

Comment (abhi18av) [reply plugins/nf-nomad/src/test/nextflow/nomad/NomadDSLSpec.groovy:150]: Mm, forgot to test that after the refactor and yes this works 👍

File: plugins/nf-nomad/src/main/nextflow/nomad/executor/NomadService.groovy (hunk 3 above)
Comment (abhi18av) [reply plugins/nf-nomad/src/main/nextflow/nomad/executor/NomadService.groovy:87]: Hi @jagedn , okay so I've tried to refactor the code keeping different layers separate. Let me know what you think about the current structure?


I do think that we can explore the idea regarding a specific "translation" class which  covers all aspects of `NomadJobDefinition` from a `NextflowTask`, however I suggest that we still continue the current iterations and revisit this once the plugin has stabilised in its design - what do you think?

**Review by jagedn — APPROVED** [review]

**Pull request merged**